UMAP: Custom initial layout

iosonofabio opened the pull request:
Let the user choose a custom initial layout for UMAP, especially useful for perturbative layouting.

Mentioned in igraph/python-igraph#529.

TODO:
- [x] Test case
- [x] Check custom initial layout dimensions

iosonofabio commented:
Would you prefer I code a dimensional check for custom initial layouts? Something like if it's not the right number of rows/columns the function should fail early?

iosonofabio commented:
Clearly it was worth it, since it found a bug in the *only* test I had written :scream:

szhorvat commented:
This parameter is named `use_seed` in other layout functions. Can we make it consistent with them, in naming and behaviour?

ntamas commented:
I just wanted to say exactly what @szhorvat said, otherwise it seems fine to me

szhorvat commented:
One more comment:

When giving error messages about vertex coordinates, I would prefer to avoid referring to "matrices" directly. This is because the manner in which high-level interfaces expose coordinates may not be a matrix, and if it _is_ a matrix, it may be the transpose of what the C interface uses.

Can you re-write it in terms of something like "number of points" and "dimension" or "embedding dimension", or whatever representation-agnostic phrasing you prefer?

Finally, with actual matrices let's not use `3 x 4` but `3 by 4`, as `x` looks weird in some environments.

iosonofabio commented:
Thanks guys. I changed the name to `use_seed`, reordered the API function args to be consistent with other functions around, and changes the error message to not mention "matrces" anymore, but just "points in X dimensions".

I also added a test to catch a corner case for trivial graphs with a seed layout, which was segfaulting.

If it passes CI I'll merge.

szhorvat commented on src/layout/umap.c:
```diff
@@ -860,11 +865,32 @@ static igraph_error_t igraph_i_layout_umap(
-        return IGRAPH_SUCCESS;
+    if (use_seed) {
+        if((igraph_matrix_nrow(res) != no_of_nodes) || (igraph_matrix_ncol(res) != ndim)) {
+          IGRAPH_ERRORF("Seed layout should have %d points in %d dimensions, got %d points in %d dimensions.",
```
`%d` is for `int` only, but we have `igraph_integer_t`, which maps to different types on different platforms. As a rule of thumb, you should never see `%d` in igraph code. If you do, it's likely a bug.

```suggestion
          IGRAPH_ERRORF("Seed layout should have %" IGRAPH_PRId " points in %" IGRAPH_PRId " dimensions, got %" IGRAPH_PRId " points in %" IGRAPH_PRId " dimensions.",
```

This will fix the test failure.
